### DISCUSSION TITLE
Streaming/Broadcasting with background support

### DISCUSSION THREAD

**Discussion by predictron-cloud** [flutterjanus/flutter_janus_client#144] [General]
Is the current version 1.2.0 (https://github.com/meetecho/janus-gateway) supported by the flutter client?
I want to switch my app from jitsi to janus client. What are the plans for this client? e.g. will it be maintained (for example, update to the newer version of flutter or janus) in next couple of years?..,

**Comment by shivanshtalwar** [comment]
For the most part this project is complete and is fully compatible with latest versions of janus-gateway as per my knowledge of janus goes, regarding support i try to support this project exclusively whenever i get time besides its being successfully used in our org so you can drop your worries about support cheers! If you liked my work i would love and appreciate some financial support on this project through donations or sponsors

**Comment by predictron-cloud** [comment]
Thanks a lot for the reply! Could you please clarify some questions:
1. Does Janus client support streaming audio and video when the application is in background? Is there a way to implement picture-in-picture mode? In the example I see dependency for flutter_foreground_task, it is used somehow?
2. For the example project, what is the difference between google_meet and video_call? I need to implement the conference with 1 presenter that will share his video and audio and ~75 members, that will see the video, but share just their audio? It should not be p2p, so that the user should send video just to the server and other users should receive the video from server. Google_meet or video_call would be a better boilerplate for it?
3. Janus can record the stream to file. How can I turn it on and specify the name of the file using the janus client?

**Comment by shivanshtalwar** [comment]
@predictron-cloud I will answer your questions point by point
1. Yes it is doable but you need to do background support by yourself to make people life easier i already created a project called screensharing example which basically uses videoroom (many to many peer) for sharing screen of users, now in your case you just have to swap the videoroom logic to streaming plugin logic and you should be good to go here's the link to [screensharing_example](https://github.com/flutterjanus/screenshare_example) project built on top of janus_client 
2. Google meet is just an example using videoroom which is act an SFU (selectively forwarding unit) it makes a mesh so that all media is transfered to every member of the room, now for your case if you want to have broadcasting feature it is something janus is more than equipped with because of something called RTP forward
So you can create a room join the broadcaster in the room and rtp forward his stream to a streaming mountpoint which all the viewers can watch at an high quality and low latency since janus streaming plugin uses threads under the hood to do the heavy lifting of media transfer and to implement such thing i would suggest you to look at the documentation of streaming and videoroom plugin
regarding videocall its just peer-peer video similar to older days of whatsapp without group call feature 😜 so videoroom and streaming plugin is your best option for implementing what you are planning to do
3. For sure both streaming and videoroom supports recording i would suggest you to use streaming plugin api to record if only 1-many configuration is required and use videoroom api to record if many-many configuration is required
you can refer to these links below
https://janus.conf.meetecho.com/docs/videoroom.html
https://janus.conf.meetecho.com/docs/streaming.html

**Reply by predictron-cloud** [reply]
@shivanshtalwar thanks for so quick and comprehensive response.
Regarding the streaming plugin I thought that the aim of it is to stream the media (e.g. mp3) from the user's device to server (and hence other users) like in your example: https://github.com/flutterjanus/flutter_janus_client/blob/master/example/lib/typed_examples/streaming.dart . And I thought that to implement streaming I need something like: https://github.com/shakurocom/janus-plugin-rtmp ? Is there an example of streaming video from camera to server, so other people can see it as a stream? As far as I understand the streaming is just more performance and network friendly way (than videoroom) to achieve my goal? What are the other benefits of streaming over the videoroom for my case? (if any)

If I select not videoroom but streaming plugin, can the participants that don't stream their video (as only the initiator will stream his video), send the audio? So there is one guy how present something with audio and video and someone, who watch the video can "turn on his microphone" and ask something and everyone will hear it.

Is it doable with streaming plugin or just with videoroom plugin?

**Reply by shivanshtalwar** [reply]
All the interactive users are supposed to be in videoroom think of it like a podcast of x users watched by y number of users through streaming plugin that is you can individually forward streams of videoroom participant to streaming mountpoint which can be watched by anyone publicly by creating a page which has streaming plugin logic in it
Each plugin has its own purpose you cannot achieve everything using either of the plugin independently
However if you want all the participants to have interactive ability you can do so by giving viewers ability to join the room with only audio track from their microphone on demand

Regarding streaming just mp3 you need to put the mp3 on janus server filesystem in order to stream it as on-demand or live type streaming mountpoint

**Reply by shivanshtalwar** [reply]
![image](https://github.com/flutterjanus/flutter_janus_client/assets/100684549/51c9fe3f-8c92-4a27-a2b3-53f996510bac)

**Reply by shivanshtalwar** [reply]
![image](https://github.com/flutterjanus/flutter_janus_client/assets/100684549/abb35f7d-bd23-411f-aad7-789cde686f4f)
Here you can see you can send plain rtp to any external consumer which can be streaming mountpoint

**Reply by predictron-cloud** [reply]
I digged into the https://github.com/flutterjanus/flutter_janus_client/blob/master/example/lib/typed_examples/streaming.dart . As far as I understand it is used for reading the list of available streams from server and show steams (audio and video) from it. I cannot understand how can I stream the video (not from) but to the server with this plugin. Maybe with createStream, but it is not obvious how and there is no example for it.

Actually, I am kind of lost in the code... I can use just videoroom but it would be sub-efficient, you convince me that streaming will be definitely better. I have one (maybe) interesting idea.... Maybe you (or maybe you know some other developer) are interested in creating of this video streaming widget for my use case commercially.... You would have all rights for this code, so you could also add it to this repository for other guys, that have similar use case.

If so, please write to predictroncloud@gmail.com
